---
title: "Spheroid"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### This is an R Markdown document going over the code necessary to analyz the spheroid data generated during the SEEP project.

```{r cars}
#inlcude graphical deption of analysis procedure
```

## Import sparse matrices from inDrops pipeline and covert to dense matrices
#### Dense matrices are in .mtx format, similar to those used in 10X Genomics analysis and are convient for storing sparse matrices densly. They are composed of a zipped gene list, cell barcode list, and counts matrix.

```{r pressure}
#code for converting to sparse matrices
#output a summary of this matrix and files output
```

## Load dense matrices into R

```{r load matrices}
# load mtx matrices
```

## Initialize and Filter Seurat objects based on distributions of UMI, Genes, and % Mitochondrial Reads per cell

```{r filter cells}
# initialize and filter
```

## Cluster cells by gene expression

```{r cluster cells}
# cluster cells
```

## Explore highly variable genes that define each cell cluster (i.e., 'marker genes')

```{r explore markers}
# define and explore marker genes
```


